cli/command/image/build.go

```diff
@@ -227,13 +239,56 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
 		buildCtx, err = archive.TarWithOptions(contextDir, &archive.TarOptions{
 			Compression:     compression,
 			ExcludePatterns: excludes,
-			IncludeFiles:    includes,
 		})
 		if err != nil {
 			return err
 		}
 	}
 
+	// replace Dockerfile if added dynamically
+	if dockerfileCtx != nil {
+		file, err := ioutil.ReadAll(dockerfileCtx)
+		dockerfileCtx.Close()
+		if err != nil {
+			return err
+		}
+		now := time.Now()
+		hdrTmpl := &tar.Header{
+			Mode:       0600,
+			Uid:        0,
+			Gid:        0,
+			ModTime:    now,
+			Typeflag:   tar.TypeReg,
+			AccessTime: now,
+			ChangeTime: now,
+		}
+		randomName := ".dockerfile." + stringid.GenerateRandomID()[:20]
+
+		buildCtx = archive.ReplaceFileTarWrapper(buildCtx, map[string]archive.TarModifierFunc{
```

Comment:
Now that `archive.ReplaceFileTarWrapper` is more generic, would it be difficult to remove `replaceDockerfileTarWrapper` and just have a single call to `archive.ReplaceFileTarWrapper` when necessary (either trusted build, Dockerfile provided on stdin, or both)? It would be nice to simplify things by removing `replaceDockerfileTarWrapper`, which seems like basically duplicate code at this point, and it would be more efficient and easier to understand if we avoid mulitple layers of wrapping. This would also address @duglin's request to use a randomized Dockerfile name when doing the trusted-build reference rewriting.